### PR TITLE
[extractor/mlb.py] Add MLBArticle extractor

### LINKED ISSUE #3475: MLB - Article falls back to generic and fails
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.04.08** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

canada

### Description

URL: "https://www.mlb.com/news/manny-machado-robs-guillermo-heredia-reacts"

Pointing directly to M3U file does download video:
https://mlb-cuts-diamond.mlb.com/FORGE/2022/2022-04/15/151a0855-302f49fd-ab9c5d2b-csvm-diamondx64-asset_5472K.m3u8

### Verbose log

```shell
yt-dlp.py -v "https://www.mlb.com/news/manny-machado-robs-guillermo-heredia-reacts"
[debug] Command-line config: ['-v', 'https://www.mlb.com/news/manny-machado-robs-guillermo-heredia-reacts']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, err utf-8, pref cp1252
[debug] yt-dlp version 2022.04.08 [7884ade65] (zip)
[debug] Python version 3.6.6 (CPython 64bit) - Windows-10-10.0.14393-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 4.3.2-2021-02-02-full_build-www.gyan.dev, ffprobe 4.3.2-2021-02-02-full_build-www.gyan.dev, rtmpdump 2.4
[debug] Optional libraries: sqlite
[debug] Proxy map: {}
[debug] [generic] Extracting URL: https://www.mlb.com/news/manny-machado-robs-guillermo-heredia-reacts
[generic] manny-machado-robs-guillermo-heredia-reacts: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] manny-machado-robs-guillermo-heredia-reacts: Downloading webpage
[generic] manny-machado-robs-guillermo-heredia-reacts: Extracting information
[debug] Looking for video embeds
ERROR: Unsupported URL: https://www.mlb.com/news/manny-machado-robs-guillermo-heredia-reacts
Traceback (most recent call last):
  File "C:\Transmogrifier\yt-dlp.py\yt_dlp\YoutubeDL.py", line 1408, in wrapper
    return func(self, *args, **kwargs)
  File "C:\Transmogrifier\yt-dlp.py\yt_dlp\YoutubeDL.py", line 1478, in __extract_info
    ie_result = ie.extract(url)
  File "C:\Transmogrifier\yt-dlp.py\yt_dlp\extractor\common.py", line 641, in extract
    ie_result = self._real_extract(url)
  File "C:\Transmogrifier\yt-dlp.py\yt_dlp\extractor\generic.py", line 4039, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.mlb.com/news/manny-machado-robs-guillermo-heredia-reacts
```

### PR DESCRIPTION
**Pull request by HobbyistDev**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

ADD DESCRIPTION HERE

This PR add site-support for mlb.com/news/. This PR didn't include url from milb

Fixes #3475 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/mlb.py [hunk 1]
```diff
@@ -347,3 +347,72 @@ def _real_extract(self, url):
...
+        }
+
+    def _real_extract(self, url):
+        display_id = self._match_id(url)
+        webpage = self._download_webpage(url, display_id)
+        apollo_cache_json = self._search_json(r'window\.initState\s*=\s*', webpage, 'window.initState', display_id)['apolloCache']
+
+        root_query_json = apollo_cache_json.get('ROOT_QUERY')
+        content_data_id = next(
+            root_query_json[key] for key in root_query_json if key.startswith('getForgeContent'))['id']
```
Comment (pukkandan) [inline yt_dlp/extractor/mlb.py:401]:
```suggestion
        content_data_id = next(
            val for key, val in root_query_json.items() if key.startswith('getForgeContent'))['id']
```
OR
```suggestion
        content_data_id = traverse_obj(root_query_json, (lambda k, _: key.startswith(k), 'id'), get_all=False)
```

File: yt_dlp/extractor/mlb.py [hunk 2]
```diff
@@ -347,3 +347,72 @@ def _real_extract(self, url):
...
+
+        # search again in apollo cache
```
Comment (pukkandan) [inline yt_dlp/extractor/mlb.py:403]: again?

File: yt_dlp/extractor/mlb.py [hunk 3]
```diff
@@ -347,3 +347,72 @@ def _real_extract(self, url):
...
+
+        content_video_query_id_list = [
+            content_video_query_id.get('id') for content_video_query_id in content_real_info.get('parts')
+            if content_video_query_id.get('typename') == 'Video']
+
+        # search again in apollo cache
+        entries = []
+        for video_query in content_video_query_id_list:
+            video_slug = traverse_obj(apollo_cache_json, (video_query, 'slug'))
+            entries.extend(list(self.get_single_video(video_slug, display_id)))
```
Comment (pukkandan) [inline yt_dlp/extractor/mlb.py:414]: should be an InAdvancePagedList with page_size=1

File: yt_dlp/extractor/mlb.py [hunk 4]
```diff
@@ -347,3 +347,72 @@ def _real_extract(self, url):
...
+        root_query_json = apollo_cache_json.get('ROOT_QUERY')
+        content_data_id = next(
+            root_query_json[key] for key in root_query_json if key.startswith('getForgeContent'))['id']
+
+        # search again in apollo cache
+        content_real_info = apollo_cache_json[content_data_id]
+
+        content_video_query_id_list = [
+            content_video_query_id.get('id') for content_video_query_id in content_real_info.get('parts')
+            if content_video_query_id.get('typename') == 'Video']
```
Comment (pukkandan) [inline yt_dlp/extractor/mlb.py:408]:
```suggestion
        content_video_query_id_list = traverse_obj(
            content_real_info, ('parts', lambda _, v: v['typename'] == 'Video', 'id'))
```

File: yt_dlp/extractor/mlb.py [hunk 5]
```diff
@@ -347,3 +347,72 @@ def _real_extract(self, url):
...
+            f'https://www.mlb.com/data-service/en/videos/{video_slug}', display_id)
+
+        formats, subtitles = [], {}
+        for playback_url in traverse_obj(json_api_data, ('playbacks', ..., 'url')) or []:
+            if determine_ext(playback_url) == 'm3u8':
+                fmt, sub = self._extract_m3u8_formats_and_subtitles(playback_url, display_id, ext='mp4')
+                formats.extend(fmt)
+                self._merge_subtitles(sub, target=subtitles)
+            elif determine_ext(playback_url) == 'jpg':
+                continue
```
Comment (pukkandan) [inline yt_dlp/extractor/mlb.py:377]: why not add this to thumbnail?

File: yt_dlp/extractor/mlb.py [hunk 6]
```diff
@@ -347,3 +347,72 @@ def _real_extract(self, url):
...
+            else:
+                formats.append({
+                    'url': playback_url,
+                    'ext': 'mp4',
+                })
+        self._sort_formats(formats)
+        yield {
```
Comment (pukkandan) [inline yt_dlp/extractor/mlb.py:384]: why yield?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/mlb.py (hunk 5 above)
Comment (HobbyistDev) [reply yt_dlp/extractor/mlb.py:377]: For link that i used in test, this url is 'fake' ( didn't actually give thumbnail url)

File: yt_dlp/extractor/mlb.py (hunk 6 above)
Comment (HobbyistDev) [reply yt_dlp/extractor/mlb.py]: changed to `yield` in the latest commit

File: yt_dlp/extractor/mlb.py (hunk 3 above)
Comment (HobbyistDev) [reply yt_dlp/extractor/mlb.py]: Sorry, i still didn't understand how this class work. Can you tell me how its work

File: yt_dlp/extractor/mlb.py [hunk 7]
```diff
@@ -347,3 +347,69 @@ def _real_extract(self, url):
...
+        self._sort_formats(formats)
+        return {
+            'id': video_slug,
+            'title': json_api_data.get('title'),
+            'description': json_api_data.get('description'),
+            'formats': formats,
+            'subtitles': subtitles,
+            'duration': parse_duration(json_api_data.get('duration')),
+            'timestamp': parse_iso8601(json_api_data.get('timestamp'))
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/mlb.py:392]: Why is this not just being passed through to `MLBVideoIE` with a url_result?

Comment (HobbyistDev) [reply yt_dlp/extractor/mlb.py]: Implemented in f9b1885846a0942ff60dd29d915dfc135498999f

File: yt_dlp/extractor/mlb.py [hunk 8]
```diff
@@ -347,3 +347,41 @@ def _real_extract(self, url):
...
+            'description': 'md5:a19d4eb0487b2cb304e9a176f6b67676',
+            'modified_date': '20220416',
+        },
+        'playlist_count': 2,
+    }]
+
+    def _real_extract(self, url):
+        display_id = self._match_id(url)
+        webpage = self._download_webpage(url, display_id)
+        apollo_cache_json = self._search_json(r'window\.initState\s*=\s*', webpage, 'window.initState', display_id)['apolloCache']
```
Comment (pukkandan) [inline yt_dlp/extractor/mlb.py:369]:
```suggestion
        apollo_cache_json = self._search_json(r'window\.initState\s*=', webpage, 'window.initState', display_id)['apolloCache']
```

File: yt_dlp/extractor/mlb.py [hunk 9]
```diff
@@ -347,3 +347,41 @@ def _real_extract(self, url):
...
+
+        # search again in apollo cache
+        entries = []
+        for video_query in content_video_query_id_list:
+            video_slug = traverse_obj(apollo_cache_json, (video_query, 'slug'))
+            entries.append(self.url_result(f'https://www.mlb.com/video/{video_slug}', ie=MLBVideoIE))
+
+        return self.playlist_result(
+            entries, content_real_info.get('_translationId'), self._html_search_meta('og:title', webpage),
+            content_real_info.get('summary'), modified_timestamp=parse_iso8601(content_real_info.get('lastUpdatedDate')))
```
Comment (pukkandan) [inline yt_dlp/extractor/mlb.py:387]: what do u think?

```suggestion
        return self.playlist_from_matches(
            traverse_obj(content_real_info, ('parts', lambda _, v: v['typename'] == 'Video', 'id')),
            getter=lambda x: f'https://www.mlb.com/video/{apollo_cache_json[x]["slug"]}',
            ie=MLBVideoIE, playlist_id=content_real_info.get('_translationId'),
            title=self._html_search_meta('og:title', webpage),
            description=content_real_info.get('summary'),
            modified_timestamp=parse_iso8601(content_real_info.get('lastUpdatedDate'))))
```

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/mlb.py (hunk 9 above)
Comment (HobbyistDev) [reply yt_dlp/extractor/mlb.py:387]: I think this fine

**Pull request merged**